tokens/tokens.md

```diff
@@ -0,0 +1,207 @@
+
+The `meta data` is an uninterpreted string but token minters are
+encouraged to use the following json object: (TODO: Extend).  The
+intention of the meta data is to have a consensus controlled way of
+providing information usable by tools (e.g., displaying an intended
+denominator, a display name, etc).
+```
+{
+  "type" : "object",
+  "properties": { "name": {"type" : "string"},
+                }
+}
+```
+
+A contract governing the usage of tokens must have a non-empty subset
+of the following ACI:
+
+```
+ spend(recipient : address, payload : Type) : boolean
+ trade([(from : address, to: address, Option(token : address))], payload : Type) : boolean
+ mint(amount: integer) : boolean
+ burn(amount : integer) : boolean
+```
+
+The contract may contain other endpoints as well, but at least one of
+the entrypoints above must be implemented. The contract can only be
+attached at create time.
+
+If a contract is provided, any transaction (spend, trade, mint, burn)
+would call this contract and the transaction only goes through if the
+result of the corresponding contract call returns true. Any other
+transaction using the token (such as contract call) would only be
+executed if a call to spend returns true. (TODO: Decide how this plays
+with contract calls that tries to pass tokens as value, etc).
+
+Note that the `final` field in the ANT takes precedence over the
+governing contract. If the ANT is final, no minting can occur.  (TODO:
+Perhaps the contract should have a `finalize` endpoint as well?)
+
+## ANT transactions
+
+### ANT create transaction (`ant_create_tx`)
+
+The ANT create transaction takes the argument:
+- Meta data : string
+- Amount
+- Recipient
+- Final
+- Contract
+- Parent  (TODO: Hierarchical tokens?)
+
+The `amount` is the number of tokens to mint at create time. Set to
+`0` if none should be minted. This can for example be combined with
+setting the `final` argument to true, thereby immediately minting all
+tokens that will ever exist.
+
+The `recipient` is the recipient of the minted tokens in `amount`. If
+not provided, the tokens are given to the creator's account.
+
+The `parent` is a pointer to a parent token for hierarchical
+tokens. (TODO: hierarchical tokens?)
+
+### ANT finalize transaction (`ant_finalize_tx`)
+
+The ANT finalize transaction takes the arguments:
+- owner
+- ANT
+
+The finalize transaction can only be submitted by the actual owner,
+and only if the ANT is not already finalized.
+
+### ANT mint transaction (`ant_mint_tx`)
+The ANT mint transaction takes the argument:
+- owner
+- ANT
+- amount
+- recipient
+- final
+
+Only the `owner` can mint new tokens, but it can pass the minted
+tokens to a recipient. If `final` is set to true, the ANT will be
+finalized after the new tokens are minted.
+
+### ANT destroy transaction
+TODO: Should we be able to destroy an ANT that has a `total_supply` of 0?
```

Comment:
What would be the incentive for the owner to destroy the ANT? Also it seems unlikely that the total supply will ever reach 0 (case-in-point: aeternity token migration).